### DISCUSSION TITLE
Questions in packetArraived function

### DISCUSSION THREAD

**Discussion by JYBESiU** [ANLAB-KAIST/KENSv3#170] [Project Q&A 2. Transport Layer (TCP)] [answered]
I would like to ask you a question if I understand the `packetArrived` function well.

According to [Tip:-Sending-and-Receiving-a-Packet](https://github.com/ANLAB-KAIST/KENSv3/wiki/Tip:-Sending-and-Receiving-a-Packet), `packetArrived` function is called by IP layer.
I think that if the client calls `connect()`, the server will have to deal with everything in the `packetArrived` function.
But is that correct that IP layer controls the everything?

And what it meaning of **Depending on the received packet, you may have to return a blocked system call.**?

**Comment by Pusnow** [comment]
> But is that correct that IP layer controls the everything?

In KENS, the central scheduler controls everything. The IP layer only passes packets to the TCP layer.

> And what it meaning of Depending on the received packet, you may have to return a blocked system call.?

There are many cases that application is blocked and resumes upon new packet arrival.
For example, the server application is blocked when it calls `accept()` and there are no established connections. After that, when new packet arrives and makes a new connection, you need to unblock the application to handle new connection.

**Reply by JYBESiU** [reply]
Then do we have to use `sendPacket('TCP')` in `packetArrived` function to pass packets from IP layer to TCP layer?

**Reply by Pusnow** [reply, marked as answer]
No, you only need to implement TCP layer. The `packetArrived` is called after all lower layers process packets.
